LFS-200: Search bar across all entities

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-200).

The new search bar at the top right.

This PR also introduces a few other changes:
1) Changing URLs from one form/questionnaire to another through the React Router did not work due to the lack of a key on `Forms.jsx`
2) `/query` can be given a `lucene` or `quick` parameter for a lucene query or basic fulltext search
3) MaterialDashboardReact dependency has been removed from Homepage, which means that we can remove it entirely now (possibly in another commit)

OrdiNeu committed:
LFS-200: Search bar across all entities
Add search bar with full text search

OrdiNeu committed:
LFS-200: Search bar across all entities
Add full text search functionality.

OrdiNeu committed:
LFS-200: Search bar across all entities
Add a new query for quick searches, to enable aggregation.
Make search results redirect the user to the form/subject/questionnaire.

OrdiNeu committed:
LFS-200: Search bar across all entities
Allow movement between different forms and questionnaires even while
looking at a form/questionnaire page
Fix sidebar closing behaviour
Remove debug code

OrdiNeu committed:
LFS-200: Search bar across all entities
Fix popper placement by removing its restriction to the main container.
Improve error reporting

OrdiNeu committed:
LFS-200: Search bar across all entities
Fix styling when the search bar is on the sidebar

OrdiNeu committed:
LFS-200: Search bar across all entities
Code style improvements
Remove the unused MaterialDashboardReact import

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -57,19 +61,108 @@ public void init(Bindings bindings)
+            final String jcrQuery = this.request.getParameter("query");
+            final String luceneQuery = this.request.getParameter("lucene");
+            final String fullTextQuery = this.request.getParameter("fulltext");
+            final String quickQuery = this.request.getParameter("quick");
```
@sdumitriu would a modular architecture be appropriate here?

marta- replied:
FTR I created https://phenotips.atlassian.net/browse/LFS-246 ti address this.

marta- replied:
FTR I created https://phenotips.atlassian.net/browse/LFS-246 ti address this.

marta- reviewed (changes requested):
- [x] Substring search should be enabled. For example, entering "scre" should find resources containing "screening". To do so, the query string should be flanked by `*`.

- [x] Additionally, `*` and other special characters in the query string need to be escaped to prevent the retrieval of results that are unexpected by a user unfamiliar with the query language. Advanced searches that make use of the query language are out of the scope of LFS-200 and should be implemented as a separate task ([LFS-247](https://phenotips.atlassian.net/browse/LFS-247)).

- [x] The resource type should be displayed in a less technical, more user-friendly way. For example, `lfs:Form` -> `Form`, or even better `Demographics data Form`

- [ ] ~~If possible, display the context of the match in that resource. E.g., for the input query `123`, display "Family id = Fam**667**_B".~~ This is postponed, to be done as part of [LFS-254](https://phenotips.atlassian.net/browse/LFS-254).

Example:
![image](https://user-images.githubusercontent.com/651980/72170795-83117f00-339f-11ea-9901-bb3461de162c.png)

or

![image](https://user-images.githubusercontent.com/651980/72176709-f8d01780-33ac-11ea-881b-e910504ad10d.png)

Edit: **Updated mockup**:

![image (1)](https://user-images.githubusercontent.com/651980/72182490-916c9480-33b9-11ea-80a5-38925ec6b184.png)

Sample code used in the mockup for producing 3 line `ListItem`s:

```
<ListItemText
   primary={(<div><Typography variant="body2" color="textSecondary">Demographics data</Typography>1077-AEB3</div>)}
   secondary={(<div>Family id = 100-<b>255</b>7-BWX</div>)} />
</ListItem>
```

Icons by resource type as used in the mockup:
* `lfs:Form`: icon = `Description`, backgroundColor=`cornflowerblue`,
* `lfs:Questionnaire`: icon = `Assignment`, backgroundColor = `limegreen`
* `lfs:Subject`: icon = `AssignmentInd`, backgroundColor = `orange`

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -0,0 +1,201 @@
+
+  // Attempt a few different methods of getting the name of an element from <code>/query?quick</code>
+  let getElementName = (element) => {
+    return element["name"] || element["title"] || element["jcr:uuid"];
```
Subject id (if available) should be one of the options. It can be listed here already and should start working for forms when [LFS-209 Linking between forms and subjects](https://phenotips.atlassian.net/browse/LFS-209) gets implemented.

OrdiNeu committed:
LFS-200: Search bar across all entities
Prepend and append wildcards to user input search

OrdiNeu committed:
LFS-200: Search bar across all entities
Add support for escaping queries

OrdiNeu committed:
LFS-200: Search bar across all entities
Add coloured avatars to search results, make the results easier to read.

OrdiNeu committed:
LFS-200: Search bar across all entities
Prepend the name of the questionnaire a form belongs to

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/SearchBar.jsx:
```diff
@@ -59,7 +63,7 @@ function SearchBar(props) {
     // The list of characters to escape are taken from https://lucene.apache.org/core/2_9_4/queryparsersyntax.html#Escaping%20Special%20Characters
     // The single quote is escaped via our QueryBuilder, so we ignore it here
-    doNotEscapeQuery ? query : query.replace("([+-!(){}[]^\"~*?:\\&& ||])", "\\$1")
+    doNotEscapeQuery ? query : (query && query.replace("([+-!(){}[]^\"~*?:\\&& ||])", "\\$1"))
```
Please do all escapting in the backend. Lucene syntax should be escaped only for the lucene search.

sdumitriu reviewed (changes requested):
- [x] There should be a limit on the number of results set. Simply focusing on the search bar returns ALL records, which takes many seconds and consumes many resources when all of Tissue Metrix is imported
- [ ] As stated above, simply focusing on the search bar already triggers a request for everything. Do we want that? As it is, no, but it may become useful if we instead order by [frecency](https://en.wikipedia.org/wiki/Frecency) (how often a result was accessed, and how recent) - [LFS-262](https://phenotips.atlassian.net/browse/LFS-262).
- [x] But frecency isn't easy to implement, so I think for this version an empty search bar should not trigger a search.
- [x] There should be a request numbering mechanism. Focusing the search bar, then typing something (in a large DB) will first display the fewer results matching the search term, but they will quickly be replaced by the irrelevant "all records" response
- [ ] Can we expand the search bar on focus? Or at least the search results? At least on my screen, the results are too narrow and don't fit properly. (For a quick example of what I mean by expand, focus the github search bar in the upper left corner) - [LFS-263](https://phenotips.atlassian.net/browse/LFS-263)
- [ ] I really like being able to open things in new tabs. Can we make the search results proper links that can be middle-/ctrl-clicked? Or is that too hard to do with Router Links? - [LFS-264](https://phenotips.atlassian.net/browse/LFS-264)

OrdiNeu committed:
LFS-200: Search bar across all entities
Refactor the results of queryBuilder to be like pagination, wherein
results can be given an offset and limit. All components using query
have been refactored to use this new interface.

SearchBar also no longer searches when no input is given, and ensures
the request ID matches.

OrdiNeu committed:
LFS-200: Search bar across all entities
Removing logging

sdumitriu reviewed (commented):
Works fine for simple text search, however it fails when entering more complicated text with special chars.

For example, `p.Met237Ile`, a protein name, doesn't bring any results. Manually searching in the node browser (and debugging in Luke) shows that:

- `select * from [lfs:Form] where contains(*, '*p\.Met237Ile*')`, as the query is executed by the current code, is interpreted as a digraph automaton trying to (correctly) match that exact substring, gives 0 matches
- `select * from [lfs:Form] where contains(*, 'p\.Met237Ile*')` (without the leading `*`), is interpreted as an exact prefix search, still 0 matches
- `select * from [lfs:Form] where contains(*, 'p\.Met237Ile')` (without the surrounding `*`), is interpreted as `+fulltext:p +fulltext:Met237Ile`: the dot is ignored, and the two parts are searched as two distinct words to match; gives one result, as expected

The main cause of the mismatch seems to be the fact that the fulltext is always analyzed, which means that in the index we don't have `p.Met237Ile`, but `p Met237Ile` as two distinct words. Trying to match the dot will always fail, since the index has no dots or other special symbols.

The proper fix is, again, to write our own custom analyzer, which would allow us to have better indexes allowing proper match of this kind of special "words".

A quick fix is to first search without wildcards, and if too few results are found, search again with wildcards and combine the two result sets.

https://phenotips.atlassian.net/browse/LFS-259

sashaandjic commented:
It would be nice if the search results are sorted the same way as the dashboard: Demographics data, Tissue data, Tumors data
![image](https://user-images.githubusercontent.com/23243244/72373704-dacd2480-36d6-11ea-9cf1-1b0d0a65201c.png)

marta- commented:
> It would be nice if the search results are sorted the same way as the dashboard: Demographics data, Tissue data, Tumors data

@sashaandjic please add this to the backlog as a separate improvement.
